compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt

```diff
@@ -0,0 +1,158 @@
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.buildField
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetObjectValueImpl
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPackageFragment
+import org.jetbrains.kotlin.ir.util.patchDeclarationParents
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformer
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+internal val assertionPhase = makeIrFilePhase(
+    ::AssertionLowering,
+    name = "Assertion",
+    description = "Lower assert calls depending on the assertions mode"
+)
+
+private class AssertionLowering(private val context: JvmBackendContext) :
+    FileLoweringPass,
+    IrElementTransformer<AssertionLowering.ClassInfo?>
+{
+    // Keeps track of the $assertionsDisabled field, which we generate lazily for classes containing
+    // assertions when compiled with -Xassertions=jvm.
+    private class ClassInfo(val irClass: IrClass, var assertionsDisabledField: IrField? = null)
+
+    override fun lower(irFile: IrFile) {
+        // In legacy mode we treat assertions as inline function calls
+        if (context.state.assertionsMode != JVMAssertionsMode.LEGACY)
+            irFile.transformChildren(this, null)
+    }
+
+    override fun visitClass(declaration: IrClass, data: ClassInfo?): IrStatement {
+        val info = ClassInfo(declaration)
+
+        super.visitClass(declaration, info)
+
+        // Note that it's necessary to add this member at the end of the class, behind all user-visible initializers,
+        // since the Java language specification demands that assertions should run during class initialization.
+        // That is also the reason why this field records whether assertions are disabled rather than enabled.
+        // During initialization, $assertionsDisabled is going to be false, meaning that assertions are checked.
+        info.assertionsDisabledField?.let {
+            declaration.addMember(it)
+
+            // Some parents of local declarations are not updated during ad-hoc inlining
+            // TODO: Remove when generic inliner is used
+            declaration.patchDeclarationParents(declaration.parent)
+        }
+
+        return declaration
+    }
+
+    override fun visitCall(expression: IrCall, data: ClassInfo?): IrElement {
+        val function = expression.symbol.owner
+        if (!function.isAssert)
+            return super.visitCall(expression, data)
+
+        val mode = context.state.assertionsMode
+        if (mode == JVMAssertionsMode.ALWAYS_DISABLE)
+            return IrGetObjectValueImpl(
```

Comment:
Is it possible to return empty expression here, e.g. IrCompositeImpl?

Reply:
You're right, that's cleaner. Changed it.